File: java/src/main/java/problems/pg42862/Solution.java
package problems.pg42862;

import java.util.Arrays;

/**
 * [PG] 42862 - 체육복 https://programmers.co.kr/learn/courses/30/lessons/42862 난이도: lv1 태그: greedy
 *
 * <p>시간복잡도: O(n) 공간복잡도: O(n)
 */
public class Solution {

  public int solution(int n, int[] lost, int[] reserve) {
    final GymClass gymClass = new GymClass(n, lost, reserve);
    gymClass.leadUniforms();
    return gymClass.countParticipants();
  }

  public static class GymClass {
    private final int[] uniforms;

    public GymClass(final int n, final int[] lost, final int[] reverse) {
      this.uniforms = new int[n];
      for (int student : lost) uniforms[student - 1]--;
      for (int student : reverse) uniforms[student - 1]++;
    }

    void leadUniforms() {
      for (int i = 0; i < uniforms.length; i++) {
        if (uniforms[i] >= 0) continue;

        if (i > 0 && uniforms[i - 1] > 0) {
          uniforms[i]++;
          uniforms[i - 1]--;
        } else if (i + 1 < uniforms.length && uniforms[i + 1] > 0) {
          uniforms[i]++;
          uniforms[i + 1]--;
        }
      }
    }

    int countParticipants() {
      return (int) Arrays.stream(uniforms).filter(u -> u >= 0).count();
    }
  }
}

File: python/problems/pg42862/solution.py
"""
[PG] 42862 - 체육복
https://programmers.co.kr/learn/courses/30/lessons/42862
난이도: lv1
태그: greedy

시간복잡도: O(n)
공간복잡도: O(n)
"""


def solution(n: int, lost: list[int], reserve: list[int]) -> int:
    students = [1] * n
    real_lost = set(lost) -set(reserve)
    real_reserve = set(reserve) - set(lost)
    for s in real_lost:
        students[s - 1] = 0
    for s in real_reserve:
        students[s - 1] = 2

    for i in range(n):
        if students[i] == 0:
            if i - 1 >= 0 and students[i - 1] > 1:
                students[i - 1] -= 1
                students[i] += 1
            elif i + 1 < n and students[i + 1] > 1:
                students[i + 1] -= 1
                students[i] += 1
    return n - students.count(0)
